prometheus-smartctl-exporter: unstable-2020-11-14 -> 0.9.1

Frostman opened the pull request:
###### Description of changes

Big bump from the 2020 patched version to the recently released 0.9.1 - https://github.com/prometheus-community/smartctl_exporter/releases

Based on #199773. I've fully tested it on my env with and without the devices specified. As a followup to @mweinelt comment https://github.com/NixOS/nixpkgs/pull/199773#issuecomment-1306243317 - here is an upstream fix you've faced https://github.com/prometheus-community/smartctl_exporter/pull/93

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Frostman committed:
prometheus-smartctl-exporter: unstable-2020-11-14 -> 0.9.1

All patches were already merged in upstream and part of the release.

Frostman committed:
prometheus-smartctl-exporter: add ldflags to set version

Frostman committed:
prometheus-smartctl-exporter: add passthru.tests

Frostman committed:
prometheus-smartctl-exporter: add Frostman to maintainers

Frostman committed:
nixos/prometheus.exporters.smartctl: use cli args instead of config

Config support was dropped and replaced with CLI args

Frostman commented:
Result of `nixpkgs-review pr 205123` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>prometheus-smartctl-exporter</li>
  </ul>
</details>

mweinelt commented on nixos/modules/services/monitoring/prometheus/exporters/smartctl.nix:
```diff
@@ -60,7 +51,7 @@ in {
       );
       ExecStart = ''
-        ${pkgs.prometheus-smartctl-exporter}/bin/smartctl_exporter -config ${configFile}
+        ${pkgs.prometheus-smartctl-exporter}/bin/smartctl_exporter --web.listen-address="${cfg.listenAddress}:${toString cfg.port}" --smartctl.path="${pkgs.smartmontools}/bin/smartctl" --smartctl.interval="${cfg.maxInterval}" ${deviceArgs}
```
We could probably shorten this by using `systemd.services.<name.>path`.

Frostman replied:
It defaults to `/usr/sbin/smartctl`, so, I think it'll not work https://github.com/prometheus-community/smartctl_exporter/blob/61e28093f6d93daefa71cc3c1b5869b4a07b6eda/main.go#L65

mweinelt replied:
I was wondering if it would pick it up from PATH when we just pass `--smartctl.path=smartctl`.

Frostman replied:
Seems like it's impossible to set a path here b/c of the exporter module structure...

mweinelt reviewed (commented):
Thanks, that's looking great, and it works on my desktop where the previous PR did not.

There is still the issue that auto-discovery fails to interact with bare NVMe devices, but you seem to have already found the udev rule to resolve that. So I wonder whether we want to include it in the module.

mweinelt commented:
Even after changing the group of the bare NVMe devices to disk, the exporter does not have permission to check the automatically discovered devices.

```
crw------- 1 root disk 247, 0 12. Nov 16:13 /dev/nvme0
crw------- 1 root disk 247, 1 12. Nov 16:13 /dev/nvme1
```

```
smartctl_exporter[1879799]: [Warning] S.M.A.R.T. output reading error: exit status 2
smartctl_exporter[1879799]: [Warning] The device error log contains records of errors.
smartctl_exporter[1879799]: [Error] Smartctl open device: /dev/nvme0 failed: Permission denied
smartctl_exporter[1879799]: [Error] smartctl returned bad data for device /dev/nvme0
````

```
❯ stat /dev/nvme?
  File: /dev/nvme0
  Size: 0         	Blocks: 0          IO Block: 4096   character special file
Device: 0,5	Inode: 113         Links: 1     Device type: 247,0
Access: (0600/crw-------)  Uid: (    0/    root)   Gid: (    6/    disk)
Access: 2022-11-12 16:13:50.412085133 +0100
Modify: 2022-11-12 16:13:50.412085133 +0100
Change: 2022-12-08 12:11:37.188926011 +0100
 Birth: -
```

Even though that should be covered by `DeviceAllow=char-nvme`:

```
❯ cat /proc/devices | grep 247
247 nvme
```

mweinelt commented on pkgs/servers/monitoring/prometheus/smartctl-exporter/default.nix:
```diff
@@ -24,7 +24,12 @@ buildGoModule rec {
+    description = ''
+      Export smartctl statistics for Prometheus
+
+      To fix permissions for scanning NVMe devices, add a similar udev rule to <literal>services.udev.extraRules</literal>:
+      <literal>SUBSYSTEM=="nvme", KERNEL=="nvme[0-9]*", GROUP="disk"</literal>
```
I think we should just set up the rule if the exporter gets enabled, if it makes auto-discovery work properly.

mweinelt replied:
Ah, we're in that funny exporter module structure. Yeah, I'm clueless as well.

Frostman replied:
I'll create a separate PR to discover options if you don't mind

Frostman commented:
@mweinelt, what's interesting about the udev rule - it seems like it worked for me only after the restart which sounds strange, but maybe there was something else

Frostman commented:
Experimenting now with 
```
services.udev.extraRules = mkIf "${name}" == "smartctl" ''
        SUBSYSTEM=="nvme", KERNEL=="nvme[0-9]*", GROUP="disk"
      '';
``` in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/monitoring/prometheus/exporters.nix#L182

Frostman commented:
Result of `nixpkgs-review pr 205123` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>prometheus-smartctl-exporter</li>
  </ul>
</details>

mweinelt commented:
Huh. Now it stopped working for me again and I don't get why.

```
ExecStart=/nix/store/2z8xfw2kz732kmz8halkmc8fgln3h93s-smartctl_exporter-0.9.1/bin/smartctl_exporter --web.listen-address="0.0.0.0:9633" --smartctl.path="/nix/store/d0v92myqqng697x4r6fdsvf6rcq930xz-smartmontools-7.3/bin/smartctl" --smartctl.interval="60s" --smartctl.device=/dev/nvme0n1 --smartctl.device=/dev/nvme1n1
```

```
❯ ls -la /dev/nvme?n?
brw-rw---- 1 root disk 259, 0  6. Dez 23:05 /dev/nvme0n1
brw-rw---- 1 root disk 259, 2  6. Dez 20:53 /dev/nvme1n1
```

```
Dez 08 19:53:20 gaia systemd[1]: Started prometheus-smartctl-exporter.service.
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.948Z caller=main.go:90 level=info msg="Starting smartctl_exporter" version="(version=0.9.1, branch=, revision=)"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.948Z caller=main.go:91 level=info msg="Build context" build_context="(go=go1.19.3, user=, date=)"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.958Z caller=main.go:112 level=warn msg="Device unavailable" name=/dev/nvme0n1
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.958Z caller=main.go:112 level=warn msg="Device unavailable" name=/dev/nvme1n1
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.958Z caller=main.go:119 level=info msg="No devices specified, trying to load them automatically"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.967Z caller=readjson.go:69 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.967Z caller=readjson.go:123 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.967Z caller=readjson.go:155 level=error msg="Smartctl open device: /dev/nvme0 failed: Permission denied"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.977Z caller=readjson.go:69 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.977Z caller=readjson.go:123 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.977Z caller=readjson.go:155 level=error msg="Smartctl open device: /dev/nvme1 failed: Permission denied"
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.977Z caller=tls_config.go:232 level=info msg="Listening on" address=[::]:9633
Dez 08 19:53:20 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:20.977Z caller=tls_config.go:235 level=info msg="TLS is disabled." http2=false address=[::]:9633
Dez 08 19:53:26 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:26.776Z caller=readjson.go:69 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
Dez 08 19:53:26 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:26.776Z caller=readjson.go:123 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
Dez 08 19:53:26 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:26.776Z caller=readjson.go:155 level=error msg="Smartctl open device: /dev/nvme0 failed: Permission denied"
Dez 08 19:53:26 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:26.793Z caller=readjson.go:69 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
Dez 08 19:53:26 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:26.793Z caller=readjson.go:123 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
Dez 08 19:53:26 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:26.793Z caller=readjson.go:155 level=error msg="Smartctl open device: /dev/nvme1 failed: Permission denied"
Dez 08 19:53:31 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:31.034Z caller=readjson.go:69 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
Dez 08 19:53:31 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:31.034Z caller=readjson.go:123 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
Dez 08 19:53:31 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:31.034Z caller=readjson.go:155 level=error msg="Smartctl open device: /dev/nvme0 failed: Permission denied"
Dez 08 19:53:31 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:31.044Z caller=readjson.go:69 level=warn msg="S.M.A.R.T. output reading" err="exit status 2"
Dez 08 19:53:31 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:31.044Z caller=readjson.go:123 level=error msg="Device open failed, device did not return an IDENTIFY DEVICE structure, or device is in a low-power mode"
Dez 08 19:53:31 gaia smartctl_exporter[2336395]: ts=2022-12-08T18:53:31.044Z caller=readjson.go:155 level=error msg="Smartctl open device: /dev/nvme1 failed: Permission denied"

Frostman commented:
Hmm, interesting, works fine for me using `systemd-run` (b/c I don't know how to safely apply it to my system which is running stable).

```
 ➜  ls -la /dev/nvme?n?
brw-rw---- 1 root disk 259, 0 Dec  8 00:04 /dev/nvme0n1
brw-rw---- 1 root disk 259, 5 Dec  8 00:04 /dev/nvme1n1
brw-rw---- 1 root disk 259, 8 Dec  8 00:04 /dev/nvme2n1
brw-rw---- 1 root disk 259, 4 Dec  8 00:04 /dev/nvme3n1
brw-rw---- 1 root disk 259, 9 Dec  8 00:04 /dev/nvme4n1
```

no errors

```
$ systemd-run -p DynamicUser=true -p SupplementaryGroups=disk -p AmbientCapabilities=CAP_SYS_ADMIN -t /nix/store/dr0xh5gk5j91p9f9hkysycmym0lbyn6y-smartctl_exporter-0.9.1/bin/smartctl_exporter --web.listen-address=127.0.0.1:9633 --smartctl.path=/nix/store/ynskqlb13y7ixk00zdjzsy1yv36db69n-smartmontools-7.3/bin/smartctl
==== AUTHENTICATING FOR org.freedesktop.systemd1.manage-units ====
Authentication is required to manage system services or other units.
Authenticating as: frostman
Password:
==== AUTHENTICATION COMPLETE ====
Running as unit: run-u1492.service
Press ^] three times within 1s to disconnect TTY.
ts=2022-12-08T19:13:29.390Z caller=main.go:90 level=info msg="Starting smartctl_exporter" version="(version=0.9.1, branch=, revision=)"
ts=2022-12-08T19:13:29.390Z caller=main.go:91 level=info msg="Build context" build_context="(go=go1.19.3, user=, date=)"
ts=2022-12-08T19:13:29.406Z caller=main.go:119 level=info msg="No devices specified, trying to load them automatically"
ts=2022-12-08T19:13:29.697Z caller=tls_config.go:232 level=info msg="Listening on" address=127.0.0.1:9633
ts=2022-12-08T19:13:29.697Z caller=tls_config.go:235 level=info msg="TLS is disabled." http2=false address=127.0.0.1:9633
```

```
 ➜  curl http://127.0.0.1:9633/metrics
...
# HELP smartctl_device_status Device status
# TYPE smartctl_device_status gauge
smartctl_device_status{device="nvme0"} 1
smartctl_device_status{device="nvme1"} 1
smartctl_device_status{device="nvme2"} 1
smartctl_device_status{device="nvme3"} 1
smartctl_device_status{device="nvme4"} 1
...
```

and the similar if running with `--smartctl.device=/dev/nvme0 --smartctl.device=/dev/nvme1`:

```
 ➜  curl http://127.0.0.1:9633/metrics
...
# HELP smartctl_device_smart_status General smart status
# TYPE smartctl_device_smart_status gauge
smartctl_device_smart_status{device="nvme0"} 1
smartctl_device_smart_status{device="nvme1"} 1
...
```

mweinelt commented:
This is somehow getting stuck in the explicit `DeviceAllow` list for me. I configure `/dev/nvme0n1` but smartctl wants to access `/dev/nvme0`. Loosening the restriction a bit makes it work.

mweinelt committed:
nixos/prometheus.exporters.smartctl: Simplify DeviceAllow logic

Setting up the DeviceAllow list with explicitly configured devices was a
nice idea, but sometimes a configured device (`/dev/nvme0n1` an NVMe
namespace) has a parent device (`/dev/nvme0`) that smartctl needs to
access to query metrics.

Falling back to the block and character definitions is probably a valid
fallback.

mweinelt commented:
The commits touching the module should probably start with `nixos/prometheus.exporters.smartctl`, not the package attribute. Other than that, I think we're good to go.

mweinelt committed:
nixos/tests/prometheus-exporters.smartctl: update test script

We can't actually get metrics for a virtual disk drive so the exporter
fails to start with 0.9.x.

Instead let's just make sure it said that /dev/vda was unavailable.
